### PR TITLE
Optimize for angle

### PR DESCRIPTION
**Pull request by ShabbyX**
Introduces the `GLSLANG_ANGLE` macro, which similarly to `GLSLANG_WEB`, reduces the grammar to what ANGLE's Vulkan backend needs, as well as stripping unused functionality. In particular, the change to `Initialize.cpp` significantly reduces the warmup time of glslang in `SetupBuiltinSymbolTable`.

### CODE REVIEW

**Comment by ShabbyX** [comment]
Note: there shouldn't be any functional changes to `glslang.m4` and `Scan.cpp`. I moved some rules and code around so I can #ifdef them out more easily.

**Comment by johnkslang** [comment]
Who will be responsible for ongoing maintenance of the accurate usage of GLSLANG_ANGLE?

**Comment by johnkslang** [comment]
I'm wondering (it will take some time) how big the overlaps are, and whether a single macro, rather than two, would suffice.

Did you notice what the big functionality gaps were between ANGLE and WEB?

One thing I imagine is that ANGLE is caring more about start up time, and WEB was caring more about footprint. (?)

**Comment by ShabbyX** [comment]
> Who will be responsible for ongoing maintenance of the accurate usage of GLSLANG_ANGLE?

The ANGLE team. We frequently roll glslang into ANGLE, and we run extensive tests (dEQP, ANGLE's end2end tests and WebGL tests). If there are regressions or build errors, we will know immediately.

**Comment by ShabbyX** [comment]
> Did you notice what the big functionality gaps were between ANGLE and WEB?

WEB uses glslang for translating gles shaders. ANGLE uses glslang for translating Vulkan GLSL features. Additionally, ANGLE supports (or will support) compute, geometry and tessellation shaders.

> One thing I imagine is that ANGLE is caring more about start up time, and WEB was caring more about footprint. (?)

ANGLE cares about both, but binary size to a lesser degree. That's why I didn't go overboard with compiling every single unnecessary line out. Where performance difference was small, I only removed larger chunks such as desktop-GL-only shader stages.

File: glslang/MachineIndependent/Initialize.cpp [hunk 1]
```diff
@@ -4603,6 +4635,7 @@ void TBuiltIns::initialize(int version, EProfile profile, const SpvVersion& spvV
     }
 
 #ifndef GLSLANG_WEB
+#ifndef GLSLANG_ANGLE
```
[the remark this answers is not in the document]
Comment (ShabbyX) [reply glslang/MachineIndependent/Initialize.cpp:4638]: As per the previous comment, ANGLE does care about binary footprint as well as performance. The latter is a more imminent issue.

**Comment by johnkslang** [comment]
@ShabbyX This pull request has some problems.

1. GitHub reported it passing all tests, but this is not true, as almost no tests actually ran.  They were later caught when tests were running again, and we traced things back to this PR.
1. The grammar does not build without generating warnings that should not be generated; this needs to be cleaned up.
1. The built-in symbol table parsing reports problems on macos, which might need addressing, if not the same issue is in the previous bullet.
1. I also saw this time (missed the first time) changing some 'int' to 'bool', making types dependent of flags, and this concerns me too.  (There was a reason to use 'int' even though it seems like it should be 'bool'.)

**Comment by ShabbyX** [comment]
@johnkslang Sorry about that. I'll focus on the first change in this PR (which optimizes time) first, which should avoid issues with 2 and hopefully 3. I'll fix whatever is the issue with 1 if it's still reproducible after removing the second commit and revert the type change per 4.

It could take a while before I can get back to this though.

**Comment by johnkslang** [comment]
I will also help by doing a more detailed review, and possibly fixing the grammar issues myself.

**Pull request merged**